test/jdk/java/lang/String/StringRacyConstructor.java

```diff
@@ -0,0 +1,418 @@
+import java.util.Arrays;
+import java.util.ConcurrentModificationException;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/*
+ * @test
+ * @bug 8311906
+ * @modules java.base/java.lang:open
+ * @summary check String's racy constructors
+ * @run junit/othervm test.java.lang.String.StringRacyConstructor
+ */
+
+public class StringRacyConstructor {
+    private static final byte LATIN1 = 0;
+    private static final byte UTF16  = 1;
+
+    private static final Field STRING_CODER_FIELD;
+    private static final Field SB_CODER_FIELD;
+
+    static {
+        try {
+            STRING_CODER_FIELD = String.class.getDeclaredField("coder");
+            STRING_CODER_FIELD.setAccessible(true);
+            SB_CODER_FIELD = Class.forName("java.lang.AbstractStringBuilder").getDeclaredField("coder");
+            SB_CODER_FIELD.setAccessible(true);
+        } catch (NoSuchFieldException ex ) {
+            throw new ExceptionInInitializerError(ex);
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // Return the coder for the String
+    private static int coder(String s) {
+        try {
+            return STRING_CODER_FIELD.getByte(s);
+        } catch (IllegalAccessException iae) {
+            throw new AssertionError(iae);
+        }
+    }
+
+    // Return the coder for the StringBuilder
+    private static int sbCoder(StringBuilder sb) {
+        try {
+            return SB_CODER_FIELD.getByte(sb);
+        } catch (IllegalAccessException iae) {
+            throw new AssertionError(iae);
+        }
+    }
+
+    // Return a summary of the internals of the String
+    // The coder and indicate if the coder matches the string contents
+    private static String inspectString(String s) {
+        try {
+            char[] chars = s.toCharArray();
+            String r = new String(chars);
+
+            boolean invalidCoder = coder(s) != coder(r);
+            String coder = STRING_CODER_FIELD.getByte(s) == 0 ? "isLatin1" : "utf16";
+            return (invalidCoder ? "INVALID CODER" : "" ) + " \"" + s + "\", coder: " + coder;
+        } catch (IllegalAccessException ex ) {
+            return "EXCEPTION: " + ex.getMessage();
+        }
+    }
+
+    /**
+     * {@return true if the coder matches the presence/lack of UTF16 characters}
+     * If it returns false, the coder and the contents have failed the precondition for string.
+     * @param orig a string
+     */
+    private static boolean validCoder(String orig) {
+        int accum = 0;
+        for (int i = 0; i < orig.length(); i++)
+            accum |= orig.charAt(i);
+        byte expectedCoder = (accum < 256) ? LATIN1 : UTF16;
```

Comment:
I think this assumes that compact strings are enabled during this test.

Reply:
Correct, the test should be enabled iff COMPACT_STRINGS is true.